Improve filter specification typings

HarelM opened the pull request:
## Launch Checklist

Improve typings for filter specifications, a PR in continue to #1383 that solves #1380.

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality.
 - [x] Document any changes to public APIs.
 - [x] Post benchmark scores.
 - [x] Manually test the debug page.
 - [x] Add an entry inside this element for inclusion in the `maplibre-gl-js` changelog: `<changelog></changelog>`.

HarelM commented:
@cns-solutions-admin feel free to review my changes.
I actually found a lot of test that covers the changed code (the else-if -> switch case).
I'm not sure how to create tests for the changed typings and how to remove the unknown, I do feel like it's better than before...

cns-solutions-admin commented on CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@
 - *...Add new stuff here...*
 - Update `icon-padding` symbol layout property to support asymmetric padding [#1289](https://github.com/maplibre/maplibre-gl-js/pull/1289)
 - Added `collaborativeGestures` option when instantiating map to prevent inadvertent scrolling/panning when navigating a page where map is embedded inline (#234)
+- Improve filter speciufication typings
```
typo. should be: improve expression and filter specification typings

cns-solutions-admin commented:
Ad testing the typings:
* write lots of expressions that should work and see if they compile
* (to test that our assumptions about what should work are correct, evaluate those expressions - the result is irrelevant, but it should not throw errors or return the default value)

HarelM commented:
The question is more about removing the `unknown` mainly as I don't feel comfortable introducing it...
Having said that, I don't have a better solution, so this will probably stay this way until, as you said, someone complains.
I also feel like in most cases people write a json style file and don't write the styles in the code so these never actually get checked (for example I use ngx-maplibre which allows writing the styles in the html file and thus doesn't get typed checked).
So in general, this is a step forward...

HarelM commented:
I tried to add the following test to the same file:
```ts
test('non-legacy exprssions', () => {
        function compileTimeCheck(filter: ExpressionFilterSpecification) {
            return isExpressionFilter(filter);
        }
        expect(compileTimeCheck(['any'])).toBeTruthy();
        expect(compileTimeCheck(['at', 2, ['array', 1, 2, 3]])).toBeTruthy();
        expect(compileTimeCheck(['at', 2, [1, 2, 3]])).toBeTruthy();
    });
```
But the test passes while the IDE (vs code) is showing the the last test it not "correct" from typescript point of view. Might be related to how jest and typescript are configured unfortunately :-(
FYI @birkskyum @wipfli

cns-solutions-admin commented:
Ad expression test: 
* ['at', 2, [1, 2, 3]] is not a valid expression, as it should be ['at', 2, ["literal", [1, 2, 3]]]
* but isExpressionFilter in reality tests, if it is not a legacy filter: "at" is not a valid legacy filter, thus it is an expression filter (albeit an invalid one)

cns-solutions-admin commented:
As the types are compile-time only, you can only check the positive case, e.g. to test for valid legacy filters:
```
    test('valid legacy filters', () => {
        function compileTimeCheck(filter: LegacyFilterSpecification) {
            expect(true).toBe(true); // not really necessary
        }
        compileTimeCheck(['has', '$id']);
    });
```

Anyway, there will probably always be expressions, that from a type perspective are correct, but due to the return types of feature properties are incorrect during evaluation.
Similar for expressions.

cns-solutions-admin commented:
BTW, above "test" is not really a test, as it is successful, if it compiles.
It can only break by not compiling.

HarelM commented:
I would like to achieve the following which at this point I don't see how:
1. Add a new expression type in the typings, for example `['foo', number]`
1. Write a test with this expression
2. Change the expression typings so that the above expression in no longer valid (i.e. remove it)
3. See that something fails this in a CI run (not in the IDE).

If you can suggest a way to accomplish this it will be great, otherwise adding these test won't protect us in the future in case we change the typings and break something...
I don't mind adding just "valid" expression tests, but I want to know that I'm protected in case I change the typings of the expression and break these tests...
As far as I tested, running `npm run test-unit` with a file that has transpilation issues doesn't fails the test or the file... :-(

cns-solutions-admin commented:
I think this is exactly, what you achieve with these "tests". They just do not fail by failing, but by not compiling.

cns-solutions-admin commented:
The problem is that jest uses @swc/jest instead of babel-jest or ts-jest for transforming the typescript code.
Babel-jest does not work, "value as ..." is invalid.
But I was successful with ts-jest:

With package.json changes:
```
    "jest": "^28.0.3",
    "jest-environment-jsdom": "28.1.3",
    "ts-jest": "^28.0.7",
```
and jest.config.js changes:
```
    'transform': {
        '^.+(feature_filter)\\.test\\.ts$': 'ts-jest',
        '^.+\\.[tj]sx?$': '@swc/jest',
    },
```
it throws compile errors, if a type test does not compile.

As a lot of tests do not compile, only the relevant ones are transformed by ts-jest.
If the tests are cleaned up to be valid typescript, all files can be compiled, but this will increase test time a lot.

HarelM commented:
The fact that a lot of tests do not compile isn't good... I'll see if I can spend some time fixing them.
Another alternative is adding a CI step that simply runs tsc on the code base in order to make sure transpilation is passing and thus avoid the penalty of running the tests slower with ts-jest (assuming this is actually faster)...

HarelM commented:
I think I found the issue talking about swc not being a type checker: https://github.com/swc-project/swc/issues/571
This might explain this issue.
I'll open a separate issue about it.

cns-solutions-admin commented:
From a testing perspective in most cases it does not really matter, if the test code is correct typescript code as long as it tests what it should test. 

However, for those tests that test typings (and those are rather few), we have to make sure that they are correct typescript code and are also type checked. This can be selectively done as described above and only marginally effects unit test speed.

HarelM commented:
I'll wait for the following PR to be merged in order to add the relevant tests and later on merge this:  #1399

HarelM commented:
I've merged the main branch after the migration to ts-node and it seems the either the test code is not accurate or there are some problems with the definition of the types that were introduced.
@cns-solutions-admin any chance you can take a look at this?
The main issue I see is with the `light.test.ts` that the test is now "not compiling", but if I change it to anything that passes compilation the test fails at runtime, so I think there's a problem with the typings...
Feel free to prove me wrong...

cns-solutions-admin commented:
Whoever wrote this test, tested that "Sh* in, sh* out", i.e. if something is passed in that cannot be understood, it passes through unchanged.

According to the documentation valid colors are strings (see https://docs.mapbox.com/mapbox-gl-js/style-spec/types/#color), e.g. '#888' (= '#888888'), '#808080', 'red'.
The result, when the value can be interpreted as color, seems to be an object {r, g, b, a}.
However, the calculation seems to have rounding problems, as '#808080' is returned as r = g = b = 0.5019607843137255 instead of r = g = b = 0.5.

So a valid test would be
```
        light.setLight({color: 'red'}, {validate: false});
        expect(light.properties.get('color')).toEqual({a: 1, r: 1, g: 0, b: 0});
```

So, the current test is bad, the typing is correct. And you can't write such a bad test in future!

P.S. unit test before: 34s, after: 238s, but second time just 70s, so not too bad.

HarelM commented:
Yea, I saw this issue too, I'm wondering if it's worth just type checking instead...

HarelM commented:
I've added a few tests, which I checked are now failing if there's a transpiration error.
I've fixed the test that was failing.
If you could write here some more valid expressions or send a PR to the relevant branch it would be great as I'm no sure I know how to cover some of the cases that you had which were problematic...

HarelM commented:
@cns-solutions-admin any chance for some help here in giving more use cases we can test before this is merged?

cns-solutions-admin commented:
I think you could add a "test" (i.e. if it compiles) for each of the expressions, but I don't think it's worth the effort.
If some expression that should work, does not work, you can still pass it as "as any as ExpressionSpecification".
And as far as I see there are no real tests for the calculation of expressions themselves, too.

HarelM commented:
I added a few tests that checks for compilation correctness as part of this PR, I think it worth the effort. Maybe not cover all cases, but at least cover some, or some complicated once at least.
You are probably correct assuming there's no full coverage for all expression types and evaluation, there are some though.
But It would still help to add more compilation tests to cover more cases and also have this as "documentation in tests" kind of thing...
You don't have to write the test, just past here a few example so I can add to the tests.
The examples you used to build the typings, assuming this is what you did...

cns-solutions-admin commented:
Examples from our configurations:
```
"line-color": ["case", ["has", "color"], ["get", "color"], "white"]
```
```
"text-field": ["case", ["all", ["has", "point_count"], ["<", ["get", "point_count"], 3]], ["get", "cluster_routes"], ""]
```
```
"circle-radius": ["case", ["has", "point_count"], ["interpolate", ["linear"], ["get", "point_count"], 2, 18.0, 10, 24.0], 12.0]
```
```
"circle-color": [
    "case",
    ["has", "point_count"], ["interpolate", ["linear"], ["get", "point_count"], 2, "#ccc", 10, "#444"],
    ["has", "priorityValue"], ["interpolate", ["linear"], ["get", "priorityValue"], 0, "#ff9", 1, "#f66"],
    "#fcaf3e"
]
```
```
"icon-image": [
    "case",
    ["==", ["get", "CAPITAL"], 1], "city-capital",
    [">=", ["get", "POPULATION"], 1000000], "city-1M",
    [">=", ["get", "POPULATION"], 500000], "city-500k",
    [">=", ["get", "POPULATION"], 100000], "city-100k",
    "city"
]
```

Filter:
```
"filter": ["match", ["get", "TYPE"], ["TARGETPOINT:HOSPITAL"], true, false]
```
```
"filter": ["match", ["get", "TYPE"], ["ADIZ", "AMA", "AWY", "CLASS", "NO-FIR", "OCA", "OTA", "P", "RAS", "RCA", "UTA", "UTA-P"], true, false]
```
```
["==", ["get", "MILITARYAIRPORT"], 1]
```

HarelM commented:
Both interpolation does not pass "compilation" from what I understand. I'm not sure if the types are incorrect or the expression... :-(
![image](https://user-images.githubusercontent.com/3269297/182844687-05610c69-0eae-4a78-9204-a8c89b847ef0.png)

cns-solutions-admin commented:
There is a missing ] for 'within' and an additional ] for interpolate.
Correct is:
```
    | ['within', unknown | ExpressionSpecification]
    // Ramps, scales, curves
    | ['interpolate', InterpolationSpecification, 
        number | ExpressionSpecification, number | ExpressionSpecification, ExpressionInputType | ExpressionSpecification, 
        ...(number | ExpressionInputType | ExpressionSpecification)[]]
```

HarelM commented:
And that's why you need tests :-)
